lib/asyn-ares.c

```diff
@@ -119,16 +119,36 @@ void Curl_resolver_global_cleanup(void)
 #endif
 }
 
+
+static void Curl_ares_sock_state_cb(void *data, ares_socket_t socket_fd,
+                                    int readable, int writable)
+{
+  struct Curl_easy *easy = data;
+  if(!readable && !writable) {
+    if(easy->easy_conn) {
+      Curl_multi_closed(easy->easy_conn, socket_fd);
+    }
+    else {
+      DEBUGASSERT(easy->easy_conn);
```

Comment:
I would suggest you *either* do the if() or the assert, not both. You either allow and expect NULLs to enter there, or not. Unless you have a reason to allow NULLs, I would prefer unconditional DEBUGASSERT there. I would also suggest a DEBUGASSERT(easy) to catch any such mistakes better.

Reply:
I definitely expect easy to be != NULL, so I could add a DEBUGASSERT(easy).
For easy->easy_conn, looking at the code I don't expect it to be NULL either, but I'm not familiar enough with the code to be 100% sure that it can't happen, do you prefer I keep the if or the assert ?

Reply:
Looking closer, it seems that `easy_conn` might be NULL in some cases, if the socket for example gets closed before there's a connection setup.

But also, if that happens this code won't call ` Curl_multi_closed()` and thus the application won't be told about the closure, which seems wrong. I'm thinking `Curl_multi_closed()` perhaps should also be modified ever so slightly to take a `struct Curl_easy *data` as input argument instead of a connection. The actual function doesn't care about the connection and just wants the easy handle...

Reply:
I just pushed the change you suggested.